### PR TITLE
PDF Generation Alternative - GSoC

### LINKED ISSUE #812: Explore other alternatives to generate PDF reports
## Project Detail

CARE is a centralized capacity management and patient management system, central to the 10BedICU Project, integrating patients, doctors, hospitals, labs, specialized treatment centers, hospital administrators, and shifting control cells. Hospitals update crucial information about their assets, providing district administration with a comprehensive view of the healthcare system via smart dashboards. CARE digitizes patient records, streamlines workflows for pandemic management, and is deployed in remote areas, enabling TeleICU services for underserved citizens. It revolutionizes healthcare management, enhancing efficiency, accessibility, and patient outcomes.

## Features To Be Implemented

**Context**

We're exploring alternatives for generating PDF reports. Our current approach involves using a browser to render HTML pages as reports via django-hardcopy. 
https://github.com/coronasafe/care/blob/19b63d2f47f41d7642698e128877651c51e5d2f3/care/facility/utils/reports/discharge_summary.py#L102-L121
However, we're interested in finding a native Python package capable of generating reports, ideally with the capability to directly create PDFs from HTML/templates.

**Action Items**

- [ ] Identify a native Python package for generating PDF reports
- [ ] Add new functions to generate the reports using the new packages
- [ ] If required change the existing report template to support the format of the packages
- [ ] Compare the resource usage of the workers for each package
- [ ] Get user feedback on the new report PDFs
- [ ] Select the best alternative by comparing the feedback on PDFs and resource usage of the packages
- [ ] Update the old methods to use the new package
- [ ] Add test cases to validate the reports generated
- [ ] Cleanup Dockerfiles

## Learning Path

|  | Details |
| :---:  | :---:  |
| Complexity       | Medium |
| Required Skills | Python, Django |
| Mentors            | @nihal467 , @vigneshhari , @khavinshankar , @rithviknishad , @gigincg , @Ashesh3 , @sainak |
| Project Size      | 175 hours |

## Link to documentation for Product Set-Up

1. For setting up the frontend, please refer to the readme file available at: https://github.com/coronasafe/care_fe
2. For setting up the backend, please refer to the readme file available at: https://github.com/coronasafe/care

## Acceptance Criteria

- All the CI/CD workflows should pass.
- Code is logically structured for long-term project sustainability.
- Proper descriptions should be included for review purposes.
- QA and Code Review

## Milestone

- [ ] Get to know CARE as a product and understand the workflow.
- [ ] Understand the user-level requirements and come up with an optimal solution.
- [ ] Implement the required PDF generations (refer above action items)
- [ ] Optimise and improve the PDF generation
- [ ] Review and QA

### PR DESCRIPTION
**Pull request by praffq**
### Documentation link 
https://docs.ohc.network/docs/care/CEP/Completed/pdf-generation

### Associated Issue
- Fixes #812 

### Merge Checklist
- [x] Updated DockerFile.
- [x]  Update prod DockerFile.
- [x] Added Test to check proper installation of Typst binary.
- [x] Adding helper for pdf generation using Typst. 
- [x] Added Test to check that the PDF is generated.
- [x] Finding Library to read and test pdf content.
- [ ] [Enhancement] Improving Error handling for pdf_generation functions.
- [x] Designing Template for Typst.
- [x] Integrating Typst with Django Templates
- [x] Writing tests
- [x] Improving test by adding more sample data
 

# Proposal Preface

Care is currently using `django-hardcopy` to generate discharge reports, which utilizes Chromium Headless for rendering Django templates in Chrome and converting them to PDF. However, this method is not very efficient and could be improved by using a native Python package that can directly convert HTML to PDF without the need for browser rendering. This would result in lower resource consumption and increased efficiency.

Currently, this function is responsible for generating PDFs, and the following is the Django template used for generating reports using Tailwind CSS and Django templates.

## Drawbacks of Current Approach

- **Dependency on Chromium Headless**: Adds significant overhead and resource consumption.
- **Increased Docker Image Size**: Including Chromium Headless impacts deployment and scalability.
- **Slower and Less Efficient Rendering**: Generating PDFs through browser emulation is slower and less efficient compared to native PDF generation.
- **Higher Memory Usage**: Requires higher memory usage during PDF generation, potentially affecting performance and responsiveness.
- **Lack of Recent Maintenance**: The absence of recent maintenance and updates for `django-hardcopy`, last seen in July 2018, has raised concerns regarding possible security vulnerabilities.

# Proposed Solution

## Initial Solution Approach

Initially, I considered utilizing a Python library like `WeasyPrint` or `xhtml2pdf`. Both libraries offer Django-friendly code. Here are the respective documentation links for further reference:
- [Xhtml2pdf](https://xhtml2pdf.readthedocs.io/)
- [WeasyPrint](https://weasyprint.readthedocs.io/)

After conducting further research and receiving feedback, I came across a recent [article](https://zerodha.tech/blog/1-5-million-pdfs-in-25-minutes/) by Zerodha that discussed PDF generation. This is where I discovered Typst.

## Final Solution (Using Typst)

Initially, PDFs were generated from HTML using Puppeteer, which involved spawning headless instances of Chrome. According to the Zerodha article, their earlier tech stack, similar to ours, relied on headless Chrome for PDF rendering, which proved inefficient at scale. However, they transitioned to Typst, noting its efficiency and scalability benefits. This sparked my interest, leading to extensive research validating Typst's effectiveness and credibility through various references.

We can also refer to the [official documentation](https://typst.app/docs/), to get detailed information related to the tool.

I engaged with the Typst community and moderators on Discord, where I received predominantly positive feedback and had productive interactions. The notable advantage I found with Typst is its freshness compared to competitors like LaTeX. While it may have fewer available resources for reference, my experience within the Typst community has been positive. The community is extremely helpful and responsive, making any potential lack of resources a minor concern.

By adopting Typst for our PDF generation needs, we can significantly improve the efficiency and scalability of our discharge report generation process, ensuring lower resource consumption and enhanced performance.

## Implementation Plan

- **Step 1:** Update Docker Files to Add Typst Dependencies

- **Step 2:** Update Helper Functions for Typst and Create a Wrapper for Typst to Compile Templates

- **Step 3:** Create a Static Template for Our Report

- **Step 4:** Integrate Typst with Django Templates in Our Project

- **Step 5:** Create Tests

- **Step 6:** Remove All Previous Dependencies and Remove Chromium and django-hardcopy

- **Step 7:** Update production files for the changes 

### Step 1: Update Docker Files to Add Typst Dependencies

Since we don't have an apt installation for Typst, we can download it from the official releases according to the build we are working on. Updated Dockerfile : [dev.Dockerfile](https://github.com/ohcnetwork/care/blob/92f717eb15120cd5d73504acc52630d33dedba19/docker/dev.Dockerfile)

### Step 2: Update Helper Functions and Create Typst Wrapper

First of all we have to create a wrapper function to allow our typ binary to compile our template. The updated function - [compile_typ](https://github.com/ohcnetwork/care/blob/92f717eb15120cd5d73504acc52630d33dedba19/care/facility/utils/reports/discharge_summary.py#L202-L240) 

After creating a wrapper we can update our helper functions. Updated helper function - [generate_discharge_summary_pdf](https://github.com/ohcnetwork/care/blob/92f717eb15120cd5d73504acc52630d33dedba19/care/facility/utils/reports/discharge_summary.py#L242-L251) 

### Step 3: Create Static Template for Reports

Static Report Template progress could be seen here - [Report Template](https://typst.app/project/rBahq8CbsixUHYwri4KtmO).

Now this document just contains different components that I've used in the template.

### Step 4: Integrate Typst with Django Templates in Our Project
Updated the previous template using HTML/CSS with `Typst`. Template can be at [patient_discharge_summary_pdf_template.typ](https://github.com/ohcnetwork/care/blob/92f717eb15120cd5d73504acc52630d33dedba19/care/templates/reports/patient_discharge_summary_pdf_template.typ)

### Step 5: Create Tests
Generating `PNG` of the pdf using typst and comparing using `Pillow` library. It involves have sample png images of the pdf in `care/facility/tests/sample_reports` folder which are to be compared with the newly generated pdf pngs , if identical the test cases passes, else throws error.

To Update the sample `PNG` files, we can update the [test_compile_typ](https://github.com/ohcnetwork/care/blob/92f717eb15120cd5d73504acc52630d33dedba19/care/facility/tests/test_pdf_generation.py#L41-L100)  function by adding the below code to test function below line 59.
```python
subprocess.run(
    ["typst", "compile", "-", sample_file_path, "--format", "png"],
    input=content.encode("utf-8"),
    capture_output=True,
    check=True,
    cwd="/",
)
```
To investigate any errors, we can remove the [finally block](https://github.com/ohcnetwork/care/blob/92f717eb15120cd5d73504acc52630d33dedba19/care/facility/tests/test_pdf_generation.py#L89-L100) from our [test_compile_typ](https://github.com/ohcnetwork/care/blob/92f717eb15120cd5d73504acc52630d33dedba19/care/facility/tests/test_pdf_generation.py#L41-L100). It'll generate the `test_output{n}.png` files in `care/facility/tests/sample_reports` folder, from where you can use image diff checker to investigate the differences.  

if in future we decide to add more data to the test function and the number of pages increases, then one should also update the [number_of_pngs_generated](https://github.com/ohcnetwork/care/blob/92f717eb15120cd5d73504acc52630d33dedba19/care/facility/tests/test_pdf_generation.py#L68) number to the number of pages of pdf generated.

### Step 6: Remove All Previous Dependencies and Remove Chromium and django-hardcopy
Updated all the functions utilising the older dependencies with the newer versions and removed `django-hardcopy` from pipfile and `chromium` from docker file.

### Step 7:  Update production files for the changes
Updated prod.Dockerfile to remove older dependencies and added newer dependencies.

# Updates in the template are listed below

### Patient Detail section

- Remove `Date of Birth` field

### Admission Details Section

- Removed `Decision After Consultation` field
- Removed `Examination details and Clinical conditions` field
- Removed `From` field
- Added `Duration of Admission` field , which shows the time span the patient was admitted (discharge date - encounter date)
- Added `Admitted to` field , which shows the bed and it's type
- Added `Diagnosis at admission` field which shows List the ICD-11 in the following order - confirmed,provisional, unconfirmed, differential
- Added `Reported Allergies` field, which shows the list of allergies
- Added `Symptoms at admission` field ,which show list of all active symptoms at the time of admission

### Health Insurance Details

- kept the table as the last item below discharge details

### Treatment Summary Section

- Combined `Prescriptions` medication details into a single table with updated format
- Removed `Treatment Plan`
- Removed `General Instructions`
- Removed `Special Instructions`
- Removed `Prescription notes` it is not relevant for discharge summary

### Discharge Summary Section

- updated `Discharge Notes` to `Discharge Advice`
- Added `Discharge Prescription` table

### Others

- Removed `Symptoms` and `Diagnosis (ICD-11)` tables and `Health Status at admission` section
- Removed `Daily Round` section
- Created three new `templatetags` , one to `format prescription`, one to `format_to_sentence_case` and one to `handle empty data`
- Added conditions to update fields name according to admission status 

## Results
- Typst is 8-15 times faster in generating PDFs
- Container size decreased by 30%
- Typst is 10-15% more memory efficient
- We eliminate the overhead associated with browser-based rendering, resulting in a more efficient and scalable process.

## Links 
- **Link to PR** - https://github.com/ohcnetwork/care/pull/2132 
- **Link to Care repo** - https://github.com/ohcnetwork/care

### CODE REVIEW

File: care/templates/reports/patient_discharge_summary_pdf_template.typ [hunk 1]
```diff
@@ -0,0 +1,564 @@
...
+#align(left, text(18pt)[== Admission Details])
+#text("")
+
+#grid(
+  columns: (1.1fr, 2fr),
+  row-gutter: 1.2em,
+  align: (left),
+  [Route to Facility:], [{{consultation.get_route_to_facility_display|field_name_to_label}}],
+  [Decision after consultation:], [{{consultation.get_suggestion_display|field_name_to_label}}],
+  [Date of Admission:], [March 18, 2024, 1:06 p.m.],
```
Comment (sainak) [inline care/templates/reports/patient_discharge_summary_pdf_template.typ]:
```suggestion
  [Date of Admission:], [{{consultation.admission_date}}],
```
hardcoded string

File: docker/dev.Dockerfile
Comment (sainak) [inline docker/dev.Dockerfile:1]: remove chromium and and update the prod Dockerfile as well

Comment (praffq) [reply docker/dev.Dockerfile:1]: Should I also remove all the dependencies of `django-hardcopy` and the previous template?

Comment (sainak) [reply docker/dev.Dockerfile:1]: yeah

**Comment by nihal467** [comment]
image

@sainak @DraKen0009 the discharge summary is not generating and the test is failing in the PR as well

**Comment by nihal467** [comment]
@DraKen0009 
image

- [ ] the duration of admission is not available for live patient , so hide or show n/a incase data is not available post taking confirmation from @sainak and aparna . We can't show temp there in the pdf

image

- [ ] only admission patient are having IP number , and all other patients are having OP number, this need to be corrected in discharge summary of non-admitted patients
- [ ] For all non-admission patient they wont be having date of admission, they will be having some other relevant based on the decision after consultation, for example for domiciliary care patient it will be date of the domicilary care commencement, so change these field in discharge summary based of the decision after consultation

image

- [ ] the diagnosis in admission is kind of confusing which of the shown diagnosis are provisional, differential, confirmed or principal diagnosis. Have a opinion taken from @sainak and aparna on how to handle these data's

image

- [ ] if you dont have data available on the above fields either mark is N/A or hide it, confirm it with @sainak and aparna on the approach on this issue
- [ ] We don't have a field called reason for admission in care, can you elaborate from which field we are fetching data for this particular field

image
image

- [ ] in the new format, compared to the old format, look like you have removed the date of the birth of the patient, whether necessary approval was taken for this remove or was it a accidental change ? My suggestion would be if age only available only show age, if dob is available show both dob and age. in the discharge summary, discuss this point with @sainak  and aparna before pushing the change on this comment

image
image

- [ ] in the symptoms , the onset date and date of cure is missing, which is a relevant data of the report
- [ ] if there is no symptoms is present for the patient , it means the patient is asymptomatic so show that in the summary

image

- [ ] entire above mention screenshot fields are missing in the new discharge summary template compared to the old template, was this intentionally removed ?

image

- [ ] the general instruction is a mandatory field in the consultation page, and when comparing to the old template, it was removed from the treatment summary, was this intentionally removed ?

image

- [ ] there is no differentiation between PRN and normal prescription, was this removed intentionally ?

image

- [ ] investigation suggestions are renamed to investigation conducted was this intentional ? in my understanding all suggestions may not be conducted, recheck with @sainak and aparna


image

- [ ] daily rounds details are removed compared to the previous template , was this intentional ?

**Comment by nihal467** [comment]
@DraKen0009 before requesting another QA review, ensure the following : 

- [ ] check all the bugs and questions are answered or resolved
- [ ] The backend test is passing
- [ ] update the PR description properly, mentioning all fields existing in the new template, along with adding a comparison comment as well, mentioning which of the following new fields are added and removed compared to the old template

CC : @gigincg @sainak

**Comment by nihal467** [comment]
**Pre-Final Testing Report**

image
imageimage

- [ ] As per the format shared by aparna, the admitted-to field should only show all the types of beds the patient was admitted to during the consultation period, not just the last admitted-to bed name and type. for example in the above screenshot case, it will be " admitted to : ICU, Bed with Oxygen Support "
- [ ] Dummy bed 6 is a bed with oxygen support, but in the discharge summary, it's been mentioned as a regular bed, make sure bed types are properly fetched.

image

- [ ] As per the PR description and the format shared by aparna, the date of admission field was planned to be removed, but in the discharge summary, this field still exists @DraKen0009 was it intentionally kept, or was it a mistake?

image

- [ ] As per the shared format, the duration of admission should be just the number of days only, so remove the time-related details from the field.

image

- [ ] When the weight and height are not available, just show N/A, without the units

image

- [ ] According to the format provided, we should only display symptoms present at the time of admission. However, in the screenshot mentioned, the first two symptoms were added during the consultation creation, and the remaining three were added while creating a log update. Therefore, instead of showing just the two symptoms, the discharge summary displays all five symptoms. Please confirm whether this was intentional with @sainak & aparna.

image
image

- [ ] As per the shared format, it was expected to be the last item after the discharge details, but the discharge summary format, you kept it on top of the discharge details , was it intentional ?

image
image

- [ ] as per the new format, you forgot to rename prescription medication to Medication administered 
- [ ] ~~notes of all the PRN, prescription and titrated medicine are missing~~ : not needed on discharge summary
- [ ] ~~in PRN prescription when the dosage is 12 mg and maximum dosage in 24 hours is 4 mg, can you confirm with @sainak and aparna that like the fourth medicine in the above screenshot we can write as 12 mg to 4 mg~~ : not a valid situation 

image
image

- [ ] as per the format rename from discharge prescription medication to discharge prescription
- [ ] the treating physician name and discharge prescription table have space between them, as per the shared format, so keep them

**Comment by nihal467** [comment]
@DraKen0009 @sainak 

- [ ] there are still bugs in the PR and few questions, please fix the bugs and comment on the questions as well. 
- [ ] Before requesting the next QA review, ensure that everything is fully functional, including all formats for both admission and non-admission, and have the functionality tested by yourself and @sainak

**Comment by praffq** [comment]
1. As per the format shared by aparna, the admitted-to field should only show all the types of beds the patient was admitted to during the consultation period, not just the last admitted-to bed name and type. for example in the above screenshot case, it will be " admitted to : ICU, Bed with Oxygen Support "
> Fixed it and also added a template tag to format the text in more readable form

2. Dummy bed 6 is a bed with oxygen support, but in the discharge summary, it's been mentioned as a regular bed, make sure bed types are properly fetched.
> Fixed 

3. As per the PR description and the format shared by aparna, the date of admission field was planned to be removed, but in the discharge summary, this field still exists DraKen0009 was it intentionally kept, or was it a mistake?
> It's intentional, asked to keep it. Have updated the PR description accordingly

4. As per the shared format, the duration of admission should be just the number of days only, so remove the time-related details from the field.
> Updated to show only days if duration is `>24 hours` else it'll show time in hours instead of 0 days

5. When the weight and height are not available, just show N/A, without the units
> Fixed 

6. According to the format provided, we should only display symptoms present at the time of admission. However, in the screenshot mentioned, the first two symptoms were added during the consultation creation, and the remaining three were added while creating a log update. Therefore, instead of showing just the two symptoms, the discharge summary displays all five symptoms. Please confirm whether this was intentional with sainak & aparna.
> Updated it. Now only the symptoms with onset date before admission date will be visible in the list

7.  As per the shared format, it was expected to be the last item after the discharge details, but the discharge summary format, you kept it on top of the discharge details , was it intentional ?
> Fixed it's kept it below the `Discharge Summary` and above `Annexure(files details)`

8. As per the new format, you forgot to rename prescription medication to Medication administered
> Fixed 

9.  as per the format rename from discharge prescription medication to discharge prescription
> Fixed

10.  the treating physician name and discharge prescription table have space between them, as per the shared format, so keep them
> Added a line space above `Treating Physician`

@nihal467  @sainak

File: care/facility/utils/reports/sample1.png
Comment (sainak) [inline care/facility/utils/reports/sample1.png:1]: move these images to facility/tests/sample_reports/.

**Review by sainak — APPROVED** [review]

**Comment by nihal467** [comment]
@DraKen0009 
image
image

- [ ] If the patient is <1 age, then show the month and days not just the month

image

- [ ] if the patient was admitted to icu bed twice during the consultation , show that only once , not twice

**Review by sainak — APPROVED** [review]

**Comment by nihal467** [comment]
LGTM @DraKen0009 nice work

**Review by nihal467 — APPROVED** [review]

File: docker/dev.Dockerfile [hunk 2]
```diff
@@ -1,3 +1,6 @@
+# Importing Typst official Image
+FROM ghcr.io/typst/typst:v0.11.0 as typstOfficialImage
```
Comment (vigneshhari) [inline docker/dev.Dockerfile]: Can we install this within the old image?  If its not doable now, lets create a backlog ticket for this.

**Review by vigneshhari — APPROVED** [review]

**Review by sainak — APPROVED** [review]

**Review by vigneshhari — APPROVED** [review]

**Pull request merged**